### PR TITLE
Go_router implementation

### LINKED ISSUE #8: Implement go_ruter as router solution for the app
add go_ruter as a navigation solution for the app

### PR DESCRIPTION
**Pull request by OppositeDragon**
Added go_router object
Screens for profile, organizations, teams, meetings details.
Moved AuthGate() to its own file.
404/error page added.
First attempt at a feature first folder structure. 

closes #8

### CODE REVIEW

File: lib/main.dart [hunk 1]
```diff
@@ -1,78 +1,110 @@
...
 
 import 'firebase_options.dart';
 
 void main() {
-  runApp(const MyApp());
+  runApp(MyApp());
 }
 
 class MyApp extends StatelessWidget {
-  const MyApp({Key? key}) : super(key: key);
```
Comment (nickmeinhold) [inline lib/main.dart:13]: Why did you remove const here?

File: lib/main.dart [hunk 2]
```diff
@@ -1,78 +1,110 @@
...
-  const HomePage({Key? key}) : super(key: key);
 
-  @override
-  Widget build(BuildContext context) {
-    return const Material(
-        child: Center(
-      child: Text('Hello'),
-    ));
-  }
+  final GoRouter _router = GoRouter(
```
Comment (nickmeinhold) [inline lib/main.dart:45]: Is this the guidance from go_router - keep _router as a top level variable in main.dart?  Just out of interest

File: lib/main.dart [hunk 3]
```diff
@@ -1,78 +1,110 @@
...
+    routes: [
+      GoRoute(
+        path: '/',
+        name: 'auth',
+        pageBuilder: (context, state) => MaterialPage(
+          key: state.pageKey,
+          child: const AuthGate(),
+        ),
+      ),
+      // GoRoute(
```
Comment (nickmeinhold) [inline lib/main.dart:56]: Why is this commented out?

File: lib/not-found/not_found_screen.dart [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import 'package:flutter/material.dart';
```
Comment (nickmeinhold) [inline lib/not-found/not_found_screen.dart:1]: "not-found" as a section seems out of place to me but I can PR for discussion later if I think of something

**Review by nickmeinhold — CHANGES_REQUESTED** [review]
Awesome work!!!  Some comments & questions, the main one to figure out I think is the order in the naming - I suggest home_screen rather than screen_home, but let's discuss and work it out!  🙂

File: lib/main.dart (hunk 3 above)
Comment (OppositeDragon) [reply lib/main.dart:56]: Because the AuthAate class is handling both the sign in page and the home page, we should probably discuss, because I think go_router should manage them individually, and be able to redirect to either of them, according to the auth status.

File: lib/not-found/not_found_screen.dart (hunk 4 above)
Comment (OppositeDragon) [reply lib/not-found/not_found_screen.dart:1]: it is a 404 page, and a page that will show errors related to navigation, while we build the app.  Moving it to another section is okay, I just didn't know where else to place it.

File: lib/main.dart (hunk 2 above)
Comment (OppositeDragon) [reply lib/main.dart:46]: Yes, it is done like that on all of the documentation.

File: lib/main.dart (hunk 3 above)
Comment (nickmeinhold) [reply lib/main.dart:57]: Awesome, yeah I think there are potentially two things to discuss then:

- routing vs. building the home page and auth page
- whether or not we adopt "don't commit commented code" style

Let's chat on Discord and come back

File: lib/main.dart (hunk 1 above)
Comment (nickmeinhold) [reply lib/main.dart:13]: @OppositeDragon did you see this question?

Comment (nickmeinhold) [reply lib/main.dart:13]: I imagine there was a non-const member added?  I just couldn't tell from a quick look and thought I'd be lazy and ask 🙂

File: lib/main.dart [hunk 5]
```diff
@@ -1,78 +1,111 @@
...
-        return StreamBuilder<User?>(
-          stream: FirebaseAuth.instance.authStateChanges(),
-          builder: (context, snapshot) {
-            // User is not signed in
-            if (!snapshot.hasData) {
-              return const SignInScreen(providerConfigs: [
-                EmailProviderConfiguration(),
-                GoogleProviderConfiguration(
-                  clientId:
-                      '220727339122-b3fflgie9rq6b6atu8sfml3274snrs31.apps.googleusercontent.com',
```
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply lib/main.dart:54]: Filed #13

File: lib/main.dart (hunk 3 above)
Comment (nickmeinhold) [reply lib/main.dart:57]: For reference, my suggestion is we adopt the same style as flutter/flutter, ie: 

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-abandonware

"...we delete rather than commenting out code. Commented-out code will bitrot too fast to be useful, and will confuse people maintaining the code."

Comment (nickmeinhold) [reply lib/main.dart:57]: Do we need to figure out the build widgets vs push pages question for this PR?  We don't really need to figure out the "no commented code" style choice either, but they're slightly different in the latter is a choice directly relevant to this PR.  If it feels like too much to bother with on this PR just sing out 🙂

**Review by nickmeinhold — CHANGES_REQUESTED** [review]
Nice one!  Thanks for making the changes!  

Just one question re: commented out code to figure out and one question left over from the previous review.

File: lib/main.dart (hunk 2 above)
Comment (nickmeinhold) [reply lib/main.dart:46]: > it is done like that on all of the documentation

Apparently not *all* documentation - the only place I looked was https://pub.dev/packages/go_router but there the `_router` object is a member of the App widget.

I hope it's clear I'm not interested in trying to be 'right' or make anyone else 'wrong', when we have a difference in perspective it's an opportunity to get closer to the truth!

Personally I'm interested in this question because I don't know go_router at all and would like to understand if there are best practices or advantages/disadvantages to different architectures.

Comment (nickmeinhold) [reply lib/main.dart:46]: Oh it's a member of HomePage?  I need get back to good habits and review in VS Code as well - github UI is good for commenting but reading code not so good imo... will do that now...

Comment (nickmeinhold) [reply lib/main.dart:46]: Anyway, sorry - this is probably better to take to Discord rather than take up energy in the PR... please feel free to resolve 👍

**Review by nickmeinhold — APPROVED** [review]
Sorry for holding this up - everything I brought up can be discussed as we go I think.  Thanks for the PR, awesome job!!!

File: lib/main.dart (hunk 2 above)
Comment (OppositeDragon) [reply lib/main.dart:46]: I see what you mean, the thing is I put the variable _router at the same level as build, that is how it is done on the docs, the thing is the MaterialApp is on one level under because of the future builder. I think the way it is now won't cause any problems. If you think _router should be at the same level of the MateralApp, it can be changed to that as well.

File: lib/main.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (OppositeDragon) [reply lib/main.dart:54]: > @OppositeDragon did you hardcode the clientID? I guess I must have done this. I wouldn't like to admit how many times I've suggested changes to my own code when reviewing PRs 🙂 Anyway I'll create an issue, feel free to resolve this 👍

That was there already.

File: lib/main.dart (hunk 2 above)
Comment (nickmeinhold) [reply lib/main.dart:46]: I was just hoping to discuss in order to better understand as I'm new to go_router.  I wasn't suggesting that I thought there was a problem or that something needed to change.

File: lib/main.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickmeinhold) [reply lib/main.dart:54]: 👍

File: lib/main.dart (hunk 2 above)
Comment (nickmeinhold) [reply lib/main.dart:46]: Oh just for clarity - I initially asked if keeping the _router object as a top level variable in `main.dart`, but it never was a top level variable in `main.dart` (that was my misunderstanding from not reading the code properly). So when you responded that all the docs say so and I disagreed, it was based on that initial misunderstanding.  Sorry for the confusion.
